src/main/java/com/google/sps/servlets/BooksServlet.java

```diff
@@ -0,0 +1,142 @@
+ */
+@WebServlet("/books")
+public class BooksServlet extends HttpServlet {
+
+  private final int NUM_BOOKS_PER_KEYWORD = 5;
+  private LinkedHashMap<String, List<Book>> booksMap;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> keywords = Arrays.asList(request.getParameterValues("key"));
+    booksMap = new LinkedHashMap<>();
+    keywords.forEach((keyword) -> addBooksForTerm(keyword));
+    String json = encodeBookMapAsJson();
+    response.getWriter().println(json);
+  }
+
+ /**
+  * Encodes the given term so that it can fit in a url.
+  * @return An encoded string
+  */
+  private String encodeTerm(String term) {
+    try {
+      return URLEncoder.encode(term, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  private String encodeBookMapAsJson() {
+    Gson gson = new Gson();
+    Map<String, Object> books = new HashMap<>();
+    books.put("books", booksMap);
+    String json = gson.toJson(books); 
+    return json;
+  }
+
+ /**
+  * Adds Book object to books from JSON results of Google Books query.
+  */
+  private void addBooksToMap(String jsonString, String keyTerm) {
+    JsonObject responseObject = new JsonParser().parse(jsonString).getAsJsonObject();
+    // Check to see if json has enough entries.
+    int numResults = responseObject.get("totalItems").getAsInt();
+    numResults = Math.min(numResults, NUM_BOOKS_PER_KEYWORD);
+    if (numResults <= 0) {
+      System.err.println("Error: no results found for query");
+      return;
+    }
+    JsonArray booksJsonArray = responseObject.getAsJsonArray("items");
+    List<Book> books = new ArrayList<>();
+    for (int i = 0; i < numResults; i++) {
+      JsonObject bookJson = booksJsonArray.get(i).getAsJsonObject().getAsJsonObject("volumeInfo"); 
+      String title = bookJson.get("title").getAsString();
+      String link = bookJson.get("infoLink").getAsString();
+      String image = bookJson.getAsJsonObject("imageLinks").get("thumbnail").getAsString();
+      String description = bookJson.get("description").getAsString();
+      String writer = formatWriters(bookJson.getAsJsonArray("authors"));
+      books.add(new Book.Builder(title, link).withImage(image).withDescription(description).withWriter(writer).build());
+    }
+    booksMap.put(keyTerm, books);
+  }
+
+ /**
+  * Reads in JSON from an open URL.
+  * @return JSON in String format.
+  */
+  private String readJsonFile(URL url) throws IOException {
+    Scanner jsonReader = new Scanner(url.openStream());
+    String data = "";
+    while(jsonReader.hasNext()) {
+        data += jsonReader.nextLine();
+    }
+    jsonReader.close();
+    return data;
+  }
+
+ /**
+  * Formats writers into string from JSON array.
+  * @return writers in String fomat.
+  */
+  private String formatWriters(JsonArray writersArray) {
```

Comment:
I would call this authors. I got confused and thought it meant writers in the technical sense like writing a file.